### PR TITLE
Fix device creation

### PR DESCRIPTION
**Pull request by acosferreira**
# Description
fix condition to UUID unique on devices, fix on conflict clause; Adjust tests

Fixes # ([2519](https://issues.redhat.com/browse/THEEDGE-2519))

## Type of change

What is it?

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update
- [ ] Tests update
- [ ] Refactor

# Checklist:

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I run `make pre-commit` to check fmt/vet/lint/test-no-fdo

### CODE REVIEW

File: pkg/services/devices.go [hunk 1]
```diff
@@ -879,7 +879,10 @@ func (s *DeviceService) platformInventoryCreateEventHelper(e PlatformInsightsCre
 		Name:        e.Host.Name,
 		LastSeen:    e.Host.Updated,
 	}
-	result := db.DB.Clauses(clause.OnConflict{DoNothing: true}).Create(&newDevice)
+	// result := db.DB.Debug().Clauses(clause.OnConflict{DoNothing: true}).Create(&newDevice)
```
Comment (omaciel) [inline pkg/services/devices.go:882]: Do you want to keep this commented out?

Comment (acosferreira) [reply pkg/services/devices.go:882]: I'm still testing if we should really do nothing or update the lastseen

File: pkg/services/devices.go [hunk 2]
```diff
@@ -879,7 +878,7 @@ func (s *DeviceService) platformInventoryCreateEventHelper(e PlatformInsightsCre
 		Name:        e.Host.Name,
 		LastSeen:    e.Host.Updated,
 	}
-	result := db.DB.Clauses(clause.OnConflict{DoNothing: true}).Create(&newDevice)
+	result := db.DB.Debug().Where(&models.Device{UUID: newDevice.UUID}).FirstOrCreate(&newDevice)
```
Comment (loadtheaccumulator) [inline pkg/services/devices.go:881]: Please add a comment on what this line is doing and maybe why.

**Review by loadtheaccumulator — CHANGES_REQUESTED** [review]
I just have the one request for a comment describing what that line is doing and maybe why it's important.

File: pkg/services/devices.go [hunk 3]
```diff
@@ -904,7 +906,9 @@ func (s *DeviceService) ProcessPlatformInventoryDeleteEvent(message []byte) erro
 		s.log.Debug("Skipping kafka message - Platform Insights Inventory message host id is undefined or event type is not delete")
 		return nil
 	}
-
+	s.log.WithFields(log.Fields{
```
Comment (AaronH88) [inline pkg/services/devices.go:909]: This log (if needed) should be moved down to where the device is actually deleted.

I also think this log should be at the debug level.

Comment (acosferreira) [reply pkg/services/devices.go]: @AaronH88 i've moved as suggested.
And I've added this log due to the many devices being created and deleted with seconds of difference and be able to capture the log if it happened again.

**Comment by AaronH88** [comment]
Good work with this.
Have you considered adding something to the BeforeCreate()  function to ensure all UUID's are unique and devices are not duplicated?

https://github.com/RedHatInsights/edge-api/blob/main/pkg/models/devices.go#L99

**Comment by loadtheaccumulator** [comment]
If it fixes the immediate problem, I'm good with it but I am concerned that Gorm many-to-many introduces unnecessary complexity and invites future issues for what could be a simple CRUD operation with a two-column table managing the relationship. More in the Jira. :-)

File: pkg/models/commits.go [hunk 4]
```diff
@@ -10,6 +10,8 @@ import (
 var (
 	// ErrOrgIDIsMandatory is for when orgID is not set
 	ErrOrgIDIsMandatory = errors.New("org_id is mandatory")
+	// ErrDeviceEsists is for UUID already registered on our DB
+	ErrDeviceEsists = errors.New("This UUID already exists")
```
Comment (de1987) [inline pkg/models/commits.go:14]: typo

File: pkg/services/devices_test.go [hunk 5]
```diff
@@ -663,6 +668,28 @@ var _ = Describe("DfseviceService", func() {
...
+				RHCClientID: event.Host.SystemProfile.RHCClientID,
+				OrgID:       event.Host.OrgID,
+				Name:        event.Host.Name,
+				LastSeen:    event.Host.Updated,
+			}
+			result = db.DB.Debug().Clauses(clause.OnConflict{DoNothing: true}).Create(&newDevice)
+			err := deviceService.ProcessPlatformInventoryCreateEvent(message)
+			Expect(err).To(BeNil())
+			var savedDevice models.Device
+			result := db.DB.Where(models.Device{UUID: event.Host.ID, OrgID: event.Host.OrgID}).First(&savedDevice).Unscoped()
```
Comment (loadtheaccumulator) [inline pkg/services/devices_test.go:684]: Why Unscoped() here?

File: pkg/services/devices_test.go [hunk 6]
```diff
@@ -631,30 +632,34 @@ var _ = Describe("DfseviceService", func() {
...
+			message, err = json.Marshal(event)
+			Expect(err).To(BeNil())
+		})
 		It("should create devices when no record is found", func() {
-			err := deviceService.ProcessPlatformInventoryCreateEvent(message)
+			err = deviceService.ProcessPlatformInventoryCreateEvent(message)
 			Expect(err).To(BeNil())
 			var savedDevice models.Device
-			result := db.DB.Where(models.Device{UUID: event.Host.ID, OrgID: event.Host.OrgID}).First(&savedDevice)
+			result := db.DB.Where(models.Device{UUID: event.Host.ID, OrgID: event.Host.OrgID}).First(&savedDevice).Unscoped()
```
Comment (loadtheaccumulator) [inline pkg/services/devices_test.go:662]: Why Unscoped() here

**Review by loadtheaccumulator — COMMENTED** [review]
/lgtm

**Review by loadtheaccumulator — APPROVED** [review]

File: pkg/services/updates_test.go [hunk 7]
```diff
@@ -76,9 +76,9 @@ var _ = Describe("UpdateService Basic functions", func() {
 					OrgID: org_id,
 				},
 			}
-			db.DB.Create(&updates[0])
-			db.DB.Create(&updates[1])
-			db.DB.Create(&updates[2])
+			db.DB.Debug().Omit("Devices.*").Debug().Create(&updates[0])
```
Comment (thom-at-redhat) [inline pkg/services/updates_test.go:79]: Is the second Debug() required?

**Review by thom-at-redhat — CHANGES_REQUESTED** [review]
2 comments

File: pkg/models/devices.go [hunk 8]
```diff
@@ -103,6 +103,12 @@ func (d *Device) BeforeCreate(tx *gorm.DB) error {
 		log.Error("device do not have an org_id")
 		return ErrOrgIDIsMandatory
 	}
+	var device Device
+	result := tx.First(&device, "UUID = ? ", d.UUID).Unscoped()
+	if result.RowsAffected > 0 {
+		log.Error("device already exists %i", result.RowsAffected)
```
Comment (thom-at-redhat) [inline pkg/models/devices.go:109]: log UUID?

**Review by thom-at-redhat — APPROVED** [review]

**Pull request merged**